### PR TITLE
Nested Properties

### PR DESCRIPTION
**Pull request by rlgnak**
This adds the ability for simpler to create nested objects from a query. It works by matching matching a column name to the object properties and nested object properties.
#### Example 1

Given the following classes and query result a Person and a nested Pet object are created. 

``` c#
    public class Person
    {
        public string Name { get; set; }
        public int? Age { get; set; }
        public Pet Pet { get; set; }
    }

    public class Pet
    {
        public string Name { get; set; }
        public int? Age { get; set; }
    }
```

| Name | Age | PetName | PetAge |
| --- | --- | --- | --- |
| Jim | 22 | Spot | 2 |
#### Example 2

Given the following classes and query result a Person and 2 nested Pet object are created. 

``` c#
    public class Person
    {
        public string Name { get; set; }
        public int? Age { get; set; }
        public Pet[] Pet { get; set; }
    }

    public class Pet
    {
        public string Name { get; set; }
        public int? Age { get; set; }
    }
```

| Name | Age | Pet0Name | Pet0Age | Pet1Name | Pet1Age |
| --- | --- | --- | --- | --- | --- |
| Jim | 22 | Spot | 2 | Fiddo | 3 |
#### Example 3

Given the following classes and query result a Person and a nested dynamic object are created. 

``` c#
    public class Person
    {
        public string Name { get; set; }
        public int? Age { get; set; }
        public dynamic Pet { get; set; }
    }
```

| Name | Age | PetName | PetAge |
| --- | --- | --- | --- |
| Jim | 22 | Spot | 2 |
#### Todo List
- Add more tests

### CODE REVIEW

**Comment by gregoryjscott** [comment]
This is looking good. Let me know when you're ready for another review.

**Comment by rlgnak** [comment]
I added support for nested dynamic objects and moved all the reflection stuff into the mappingObject. I also made a change to how dynamics are created you just pass dynamic as a generic type when creating a task `new BuildObject<dynamic>()` let me know what you think.

I believe I have it to a point where a code review would be helpful.

File: app/Simpler.Tests/Core/Mocks/MockComplexPerson.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+        public MockEnum MockEnum { get; set; }
+    }
+
+    public class MockPet
+    {
+        public string Name { get; set; }
+        public int? Age { get; set; }
+    }
+
+    public class MockVechile
```
Comment (gregoryjscott) [inline app/Simpler.Tests/Core/Mocks/MockComplexPerson.cs]: Typo, should be Vehicle.

File: app/Simpler.Tests/Data/Tasks/BuildComplexTypedTest.cs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+            // Assert
+            Assert.That(task.Out.Objects.Count(), Is.EqualTo(2));
+            Assert.That(task.Out.Objects[0].Pet.Name, Is.EqualTo("Fiddo"));
+            Assert.That(task.Out.Objects[1].Pet.Name, Is.EqualTo("Spot"));
+            Assert.That(task.Out.Objects[0].Vechiles[0].Make, Is.EqualTo("Dodge"));
+            Assert.That(task.Out.Objects[1].Vechiles[1].Make, Is.EqualTo("Jeep"));
+        }
+    }
+}
```
Comment (gregoryjscott) [inline app/Simpler.Tests/Data/Tasks/BuildComplexTypedTest.cs]: This needs more test cases. Examples: people with no pets, people with no cars, people with more than two cars, and all those combined.

File: app/Simpler/Data/Tasks/BuildMappings.cs [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        }
+
+        public override void Execute()
+        {
+            var root = new ObjectMappingRoot { PropertyType = In.RootType };
+            var sortedColumnNames = In.ColumnNames.OrderBy(s => s.Key);
+            foreach (var columnName in sortedColumnNames)
+            {
+                ProcessColumnName(root, columnName.Key);
+            }
```
Comment (gregoryjscott) [inline app/Simpler/Data/Tasks/BuildMappings.cs]: I see here that it's iterating each column in the root node and "processing" it, but I have no idea what that really means. It appears `ProcessColumnName` that is mutating `root`, but the name `ProcessColumnName` doesn't make that clear.

I would suggest writing some tests for `ProcessColumnName` that describe what it does and then rename it appropriately. Also, I'd usually prefer a specific `Task` for this purpose, but I don't know if using a task for recursion is a good fit.

**Comment by rlgnak** [comment]
I did some refactoring in an attempt to make the code more understandable and added more unit tests. 

I also renamed `ProcessColumnName` to `ParseColumnName` and created a set of unit tests for it.  I'm not sure that is much better, but I'm stumped on how to make it any clearer,

**Comment by gregoryjscott** [comment]
This PR will be re-created after Simpler.Data is resurrected. See #19 for details.

**Pull request closed without merging**